### DISCUSSION TITLE
Sky positions

### DISCUSSION THREAD

**Discussion by WillHartley** [skypyproject/skypy#542] [Ideas]
It would be very useful to have positional functionality in the short-ish timescale, and I think there are many uses that we need that are distinct and separate from the physical dark matter density distribution of the Universe.

Examples:

* Even random galaxy positions are useful for many mock galaxy sets.
* Survey characteristics maps (e.g. coverage maps, variation of depth across a survey - the UltraVISTA deep stripes, for example).
* Stellar density map.

To me, this feels like it's own module that would be called by a galaxies simulation (for instance), though I don't yet have a clear idea of how it should be implemented. Via Healpy / Healsparse is one thought, though survey boundaries often are expressed in the form of polygons / trivial shapes.

**Comment by rrjbca** [comment]
FYI there was previously a [skypy.positions module](https://skypy.readthedocs.io/en/v0.3/position/index.html) that could sample from a circular area around a given position on the sky, or from within a healpix pixel. This was removed before v0.4, I don't remember exactly but I believe it was shelved until we could implement a more complete scheme for simulation volumes based on healpix maps and redshift slices (which we also prototyped as ["lightcone mode"](https://skypy.readthedocs.io/en/v0.3/pipeline/index.html#lightcone-simulations)).

I could certainly look at reverting the positions module onto a branch if there was interest in it? However if we do want to go ahead with something like "lightcone mode" it probably needs a more fundamental redesign.

**Reply by WillHartley** [reply]
Thanks! Good to know that some things already exist.

I think there's a need / space for both approaches, but how it's organised into modules isn't so clear to me. Maybe skypy.positions would be so light that it would end up making more sense as just being part of skypy.lightcone, but I can imagine some complexity entering later.

Here are a couple of concrete use cases:

1) The image depth of a survey such as DES varies across the footprint, and we have maps in healpix format to describe that variation. We want to be able to include that information and its impact on the object detection (or selection) and photometric errors. Similar for seeing FWHM when it comes to image simulation (if we head down that route).

2) The XMM-LSS field contains within it deeper survey fields with additional bands. We want those extra bands only for the objects that lie within the correct area.

3) Simulating a sample that might go into a spectroscopic programme using MOONS (for example), means that we need to remove areas that are inaccessible due to bright stars - preferably the real star mask for the chosen survey areas.

All of these can be post-processed from current skypy outputs, but I think would benefit from being natively included.
Also, realistic clustering of galaxies would be a great addition, and clearly needs a lightcone approach, but I think there's a big step up in functionality of skypy with even just random galaxy positions if these sort of footprint-like quantities can be included.

So maybe resurrecting the v0.4 module is useful, but maybe better is to think more in terms of these non-galaxy position effects?

**Comment by Lucia-Fonseca** [comment]
I'd very much need some function to produce a catalog given a healpix mask: the catalog would be Object_ID, RA, DEC, redshift.
I need something very simple:
* I can draw a sample of randomly distributed positions (RA, DEC) over the sphere.
* Apply the healpix mask.
* Use the pipeline to draw from the redshift distribution.

*Inputs*:
* healpix mask
* optional: RA_min, RA_max, DEC_min, DEC_max

*Outputs*
* RA, DEC
* Or Object_ID, RA, DEC

We could either revive skypy.positions or create a skypy.survey and add positions with the random function and also the uniform_around function. 

Is this reasonable @itrharrison @rrjbca @WillHartley ?

**Reply by WillHartley** [reply]
Here's a quick and hacky notebook to show the use of the function to draw RA, Dec values within healpix pixels, as Richard mentioned the other week.

http://star.ucl.ac.uk/~whartley/distribute_among_healpix.ipynb

For a complex mix of nested NSIDE pixels, you need to get the area of each pixel to put into the multinomial weighting.

**Reply by Lucia-Fonseca** [reply]
Thanks @WillHartley would you be able to share the npz file too? thanks

**Comment by WillHartley** [comment]
Yes, of course!
[DES_DF_Ksband_maglim_map.zip](https://github.com/skypyproject/skypy/files/9084402/DES_DF_Ksband_maglim_map.zip)
